drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2494 @@
+ * Name: nvs_read_hist
+ *
+ * Description:
+ *   Read a history entry from the file system. But expired ones will return
+ *   -ENOENT.
+ *   Only when we are updating an existing kv with new data and
+ *   right before the prev one hasn't been expired, will there
+ *   be two newest entries. At that time, nvs_read_hist is used to search
+ *   for prev one.
+ *
+ * Input Parameters:
+ *   fs       - Pointer to file system
+ *   key      - Key of the entry to be read
+ *   key_size - Size of key
+ *   data     - Pointer to data buffer
+ *   len      - Number of bytes to be read
+ *   cnt      - History counter: 0: latest entry, 1:one before latest ...
+ *   ate_addr - the addr of found ate
+ *
+ * Returned Value:
+ *   Number of bytes read. On success, it will be equal to the number
+ *   of bytes requested to be read. When the return value is larger than the
+ *   number of bytes requested to read this indicates not all bytes were
+ *   read, and more data is available. On error returns -ERRNO code.
+ *
+ ****************************************************************************/
+
+static ssize_t nvs_read_hist(FAR struct nvs_fs *fs, FAR const uint8_t *key,
+                size_t key_size, FAR void *data, size_t len, uint16_t cnt,
+                FAR uint32_t *ate_addr)
+{
+  int rc;
+  uint32_t wlk_addr;
+  uint32_t rd_addr;
+  uint32_t hist_addr;
+  uint16_t cnt_his;
+  struct nvs_ate wlk_ate;
+  uint32_t hash_id;
+
+  if (len > (fs->geo.erasesize - 2 * sizeof(struct nvs_ate)))
+    {
+      ferr("Read buffer is too large\n");
+      return -EINVAL;
+    }
+
+  cnt_his = 0;
+
+  hash_id = nvs_fnv_hash(key, key_size) % 0xfffffffd + 1;
+
+#if CONFIG_MTD_CONFIG_FAIL_SAFE_LOOKUP_CACHE_SIZE > 0
+  wlk_addr = fs->lookup_cache[nvs_lookup_cache_pos(hash_id)];
+  if (wlk_addr == NVS_LOOKUP_CACHE_NO_ADDR)
+    {
+      rc = -ENOENT;
+      goto err;
+    }
+
+#else
+  wlk_addr = fs->ate_wra;
+#endif
+  rd_addr = wlk_addr;
+
+  while (cnt_his <= cnt)
+    {
+      rd_addr = wlk_addr;
+      hist_addr = wlk_addr;
+      rc = nvs_prev_ate(fs, &wlk_addr, &wlk_ate);
+      if (rc)
+        {
+          ferr("Walk to previous ate failed, rc=%d\n", rc);
+          goto err;
+        }
+
+      if ((wlk_ate.id == hash_id) && (nvs_ate_valid(fs, &wlk_ate)))
+        {
+          if ((wlk_ate.key_len == key_size)
+              && (!nvs_flash_block_cmp(fs,
+              (rd_addr & ADDR_BLOCK_MASK) + wlk_ate.offset, key, key_size)))
+            {
+              cnt_his++;
+            }
+          else
+            {
+              fwarn("hash conflict\n");
+            }
+        }
+
+      if (wlk_addr == fs->ate_wra)
+        {
+          break;
+        }
+    }
+
+  /* skip the deleted or expired entry, or gc_done_ate(id mustn't match) */
+
+  if (((wlk_addr == fs->ate_wra) && (wlk_ate.id != hash_id)) ||
+      (wlk_ate.expired != CONFIG_MTD_CONFIG_ERASEDVALUE) ||  (cnt_his < cnt))
```

Comment:
After we update a KV, we need to expire the old ate.
This is where we want to read cnt==1.

Reply:
But, the first match is old ate, why need search the next one?

Reply:
the first match is the new ate in this case. 
Because we just finished writing the new ate, but expiring is not done. Power lose happens before expiring.
Next time we bootup nvs, there are two new ate in NVS.